creator/utils.py

```diff
@@ -0,0 +1,85 @@
+import time
+import logging
+from rq.job import NoSuchJobError, JobStatus
+from rq.command import send_kill_horse_command
+from rq.worker import Worker, WorkerStatus
+
+WAIT_FOR_FAIL_TIMEOUT = 30
+
+
+def stop_job(job_id, queue=None, delete=True, logger=None):
+    """
+    Stop a running rq job and optionally delete the job from the rq queue
+    and registries.
+
+    NOTE: This method may not be needed after this
+    https://github.com/rq/rq/issues/1371 is implemented
+
+    For jobs that have failed, do nothing
+    For jobs that are queued or deferred cancel or delete them (see rq.job.Job)
+    For jobs that have started, kill the worker horse executing the job
+
+    If delete is True, remove the job from the redis queue, all rq registries,
+    and from redis
+
+    If delete is False just remove the job from the redis queue
+
+    :param job_id: The redis job id
+    :type job_id: str
+    :param queue: The queue the job was submitted to
+    :type queue: rq.Queue
+    :param delete: whether to delete or cancel the job
+    :type delete: bool
+    """
+    if not logger:
+        logger = logging.getLogger(__name__)
+
+    logger.info(f"Preparing to stop redis job {job_id}")
+
+    # Find rq job
+    if not queue:
+        queue = Queue(connection=Redis())
+
+    redis_job = None
+    try:
+        redis_job = queue.fetch_job(job_id)
+    except NoSuchJobError:
+        return
+
+    # Skip jobs that have failed or are finished
+    if (
+        (not redis_job) or
+        (redis_job.get_status() in {JobStatus.FAILED, JobStatus.FINISHED})
+    ):
+        logger.info(f"No redis job: {job_id} found, aborting")
+        return
+
+    # Kill running jobs
+    if redis_job.get_status() == JobStatus.STARTED:
+        for worker in Worker.all(queue=queue):
+            if (
+                (worker.state == WorkerStatus.BUSY) and
+                (worker.get_current_job().id == job_id)
+            ):
+                logger.info(f'Killing job {job_id}')
+                send_kill_horse_command(queue.connection, worker.name)
+
+    # Wait until job has been moved to failed job registry to delete or
+    # cancel job
+    while time.time() < (time.time() + WAIT_FOR_FAIL_TIMEOUT):
```

Comment:
How long does this operation take and where do we plan to call it from?
The `while` is going to peg the CPU, possibly of a worker that might otherwise be serving requests, which could cause our autoscaler to kick up the instance count. It will also bombard redis with requests which can be a problem if we end up canceling lots of jobs at once.
Perhaps it would be better to queue another job to check up on the status ?

Reply:
This operation will be called from the `IngestRun.cancel` method. Eh yea 30s was left from debugging. I think I'm just going to sleep for less than a second, maybe 0.1 bc I noticed during testing that's all it takes for rq to move the job to a registry. 

In general submitting another job to check on status seems like it would have the same problem bc the status might not have changed by the time you're inside the job. In this case it would probably work but I'd rather just sleep. Sigh the real solution is to have rq support subscription to job status changes. This issue looks like what we want: https://github.com/rq/rq/issues/717 but no ones looked at it in a while